src/tunnel.cc

```diff
@@ -1307,37 +1298,74 @@ TunnelStateData::saveError(ErrorState *error)
 {
     debugs(26, 4, savedError << " ? " << error);
     assert(error);
+    // nobody saves finalError (or makes finalError equal to savedError)
+    assert(!finalError || (finalError != error && finalError != savedError));
+    if (savedError == error)
+        return;
     delete savedError; // may be nil
     savedError = error;
 }
 
-/// Starts sending the given error message to the client, leading to the
-/// eventual transaction termination. Call with savedError to send savedError.
+/// Makes sure the transaction is terminated, immediately (if possible) or after
+/// reporting the previously saved error to the client and/or finishing in-progress
+/// I/Os (as needed).
+/// All callers guarantee savedError existence, which becomes the final error during
+/// the first call. Safe to call multiple times.
 void
-TunnelStateData::sendError(ErrorState *finalError, const char *reason)
+TunnelStateData::sendErrorAndDestroy(const char *reason)
 {
     debugs(26, 3, "aborting transaction for " << reason);
+    assert(savedError);
+
+    if (finalError) {
+        debugs(26, 4, "the final error already exists: " << finalError << ", discarding new error: " << savedError);
+        assert(finalError != savedError);
+        // get rid of an error created after finalError
+        delete savedError;
+        savedError = nullptr;
+        return;
+    }
+
+    finalError = savedError;
+    savedError = nullptr;
 
     if (request)
         request->hier.stopPeerClock(false);
```

Comment:
This feels premature if we are still writing to the peer... Move lower, just before deleteThis()? tunnelServerClosed() will then handle the case where we are still writing to the peer AFAICT...

Reply:
> Move lower, just before deleteThis()?

This does not cover some other cases, e.g., when we need to send the error and server connection does not exist (i.e., tunnelServerClosed() is not going to be called). I think we should call this method once in TunnelStateData destructor (done at 23d331f). I added an additional method  so that stopPeerClock() was conditional on startPeerClock().